src/Cake.Core/ICakeTaskInfo.cs

```diff
@@ -27,6 +27,12 @@ public interface ICakeTaskInfo
         /// Gets the task's dependencies.
         /// </summary>
         /// <value>The task's dependencies.</value>
-        IReadOnlyList<string> Dependencies { get; }
+        IReadOnlyList<CakeTaskDependency> Dependencies { get; }
```

Comment:
Just as a note, this change broke our build. But it was an easy fix once we figured it out.

We have a very non-standard usage where we dynamically generate CakeTask objects as a pre-build step.

Reply:
Sorry to hear that you had issues. Could you provide an example of what you are doing, and the change you had to make?

Reply:
TL;DR : 
```c#
   if(!taskB.Dependencies.Contains(taskA.Name))
```
This line broke due to the highlighted line in the diff.

Once we ruled out a couple of other possible causes, the fix was simply:
```c#
   if(!taskB.Dependencies.Select(dep => dep.Name).Contains(taskA.Name))
```
in retrospect we could have also used
```c#
   if(!taskB.Dependencies.Any(dep => dep.Name == taskA.Name))
```

**Longer version**
Without going into extreme detail, we a have bunch of builds that are structurally very similar. Rather than defining N separate cake builds, we wrote a 'meta' layer in cake that takes a couple of params and then generates the needed cake tasks at build time. It actually works really well since Cake takes care of the dependency graph ordering for us. 

The crux of the operation looked like this before the update.
```c#
delegate CakeTemplate<TBuildContext>[] BuildFlowSelector<TBuildContext>(TBuildContext context) where TBuildContext : IBuildContext;

CakeTask ComposeFlow<TBuildContext>(BuildFlowSelector<TBuildContext> selector, IEnumerable<TBuildContext> contexts)
        where TBuildContext : IBuildContext
{
    var headOfTaskChain = contexts
        .SelectMany(ctx => selector(ctx).Select(template => GenerateTask(template,ctx)))
        .Aggregate((taskA,taskB) => {
            //Information($"{taskA.Name} -> {taskB.Name}");
            if(!taskB.Dependencies.Contains(taskA.Name))
                taskB.AddDependency(taskA.Name);
            return taskB;
        });

    return headOfTaskChain;
}
```
I know that is a bit dense but the issue was on this line 

```c#
   if(!taskB.Dependencies.Contains(taskA.Name))
```

This ensures the task isn't already linked and links if needed. Note that the inferred signature of `Contains` is `bool Contains<T>(this IEnumerable<T> enumerable, T item)` where `T` is string.

After Cake 0.23 went out `taskB.Dependencies` was no longer an `IReadOnlyList<string>` but now an ` IReadOnlyList<CakeTaskDependency>` as noted in the diff above. But the second argument of `Contains`  was still `taskA.Name` which was a string. The type mismatch caused the overload resolution to fail and make a bogus suggestion about IQueryable being the closest match. Which sent me down a short goose chase before i realized exactly what was going on.  

No real harm done, just figured i would bring it up.